drivers/iio/adc/max14001.c

```
@@ -0,0 +1,359 @@
+#define MAX14001_MASK_FLAGS_CRCF		BIT(6)
+#define MAX14001_MASK_FLAGS_FET			BIT(7)
+#define MAX14001_MASK_FLAGS_MV			BIT(8)
+
+/* MAX14001_REG_FLTEN MASKS */
+#define MAX14001_MASK_FLTEN_DYEN		BIT(0)
+#define MAX14001_MASK_FLTEN_EADC		BIT(1)
+#define MAX14001_MASK_FLTEN_EINRD		BIT(2)
+#define MAX14001_MASK_FLTEN_ESPI		BIT(3)
+#define MAX14001_MASK_FLTEN_ECOM		BIT(4)
+#define MAX14001_MASK_FLTEN_ECRCL		BIT(5)
+#define MAX14001_MASK_FLTEN_ECRCF		BIT(6)
+#define MAX14001_MASK_FLTEN_EFET		BIT(7)
+#define MAX14001_MASK_FLTEN_EMV			BIT(8)
+
+/* MAX14001_REG_WEN values*/
+#define MAX14001_REG_WEN_WRITE_ENABLE	0x294
+#define MAX14001_REG_WEN_WRITE_DISABLE	0x0
+
+enum max14001_chip_model {
+	max14001,
+	max14002,
+};
+
+struct max14001_chip_info {
+	const char *name;
+	/* TODO: Add more information */
+};
+
+static struct max14001_chip_info max14001_chip_info_tbl[] = {
+	[max14001] = {
+		.name = "max14001",
+	},
+	[max14002] = {
+		.name = "max14002",
+	},
+};
+
+struct max14001_state {
+	struct spi_device *spi;
+	const struct max14001_chip_info *chip_info;
+	int vref_mv;
+};
+
+static int max14001_get_vref_mv(struct max14001_state *st)
+{
+	struct device *dev = &st->spi->dev;
+	int ret = 0;
+
+	ret = devm_regulator_get_enable_read_voltage(dev, "vrefin");
+	if (ret < 0){
+		st->vref_mv = 1250000 / 1000;
+		dev_info(&st->spi->dev, "%s: vrefin not found. vref_mv %d\n", __func__, st->vref_mv);
+	} else {
+		st->vref_mv = ret / 1000;
+		dev_info(&st->spi->dev, "%s: vrefin found. vref_mv %d\n", __func__, st->vref_mv);
+	}
+
+	return ret;
+}
+
+static int max14001_init_required_regulators(struct max14001_state *st)
+{
+	struct device *dev = &st->spi->dev;
+	int ret = 0;
+
+	ret = devm_regulator_get_enable(dev, "vdd");
+	if (ret)
+		return dev_err_probe(dev, ret, "Failed to enable specified Vdd supply\n");
+
+	ret = devm_regulator_get_enable(dev, "vddl");
+	if (ret)
+		return dev_err_probe(dev, ret, "Failed to enable specified Vddl supply\n");
```

All this could go directly into the probe function. Before reading of/device tree properties. These are required power supplies, so best to fail early if any of them is absent.

---

Done, thank you!